### PR TITLE
add support for generating Jenkinsfile & tox.ini using "python -m vsc.install.ci --force" (HPC-7603)

### PR DESCRIPTION
**Pull request by boegel**
* updated `Jenkinsfile` & added `tox.ini` included here were auto-generated with `python -m vsc.install.ci --force`
* a `revision` tag is included in both auto-generated `Jenkinsfile` and `tox.ini` files, so the tests can verify they're up-to-date (or need to be updated via `python -m vsc.install.ci --force`); I'm deliberately *not* using the `vsc-install` version in the revision tag (not all future `vsc-install` updates will result in changes to the generated `Jenkinsfile` or `tox.ini`)
* Travis CI config (`.travis/yml`) was updated (manually) to use `tox` & test with both Python 2.7 and 3.6 (which helped a lot in making sure this all worked before opening the PR and letting Jenkins run the tests)
* tests with Python 3.6 will fail for `vsc-install` in Travis due to compatibility issues that haven't been fixed yet, but `tox` is configured to allow failing tests for now
* tests with Python 3.6 will fail in Jenkins for all `vsc-*` packages for now (since the required `python3.6` is missing on `jenkins8`), but `tox` has been configured to ignore that (to avoid having to move all `vsc-*` packages to the new (not-ready-yet) Jenkins server once this gets merged); this will be fixed when moving `vsc-install` (& co) to the new Jenkins server that has both Python 2.7 and 3.6 available (not easy to do on current `jenkins8`)
* tests were added to `vsc.install.commontest` to verify that `Jenkinsfile` and `tox.ini` are present and up-to-date for all `vsc-*` packages (based on the revision tag)

Note: merging this will effectively require that all other `vsc-*` Python packages first run `python -m vsc.install.ci --force` to re-generate the `Jenkinsfile` (& `tox.ini`), and that they're added to the repository, otherwise the tests added to `commontest` will fail.

Tests will fail everywhere as soon as the new `vsc-install` is published on PyPI...

### CODE REVIEW

**Comment by boegel** [comment]
See also re-generated `Jenkinsfile` and `tox.ini` for `vsc-base` in https://github.com/hpcugent/vsc-base/pull/285

**Comment by boegel** [comment]
`tox` installation on Jenkins fails with:

```
Moving filelock-3.0.12-py2.7.egg to /var/lib/jenkins/.local/lib/python2.7/site-packages
error: [Errno 2] No such file or directory: '/var/lib/jenkins/.local/lib/python2.7/site-packages/filelock-3.0.12-py2.7.egg'
```

I'll figure that one out later...

File: Jenkinsfile [hunk 1]
```diff
@@ -1,6 +1,13 @@
...
 node {
-  stage 'checkout git'
-  checkout scm
-  stage 'test'
-  sh "python setup.py test"
-}
+    stage 'checkout git'
+    checkout scm
+    stage 'test'
+    sh 'python2.7 -V'
```
Comment (stdweird) [inline Jenkinsfile]: what is this for?

File: Jenkinsfile [hunk 2]
```diff
@@ -1,6 +1,13 @@
...
+    sh 'python -m easy_install -U --user tox'
+    sh 'tox -v'
+}
```
Comment (stdweird) [inline Jenkinsfile]: add newline

File: lib/vsc/install/ci.py [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+"""
+from __future__ import print_function
+from optparse import OptionParser
+import logging
+import os
+import sys
+
+
+JENKINSFILE = 'Jenkinsfile'
+JENKINSFILE_REVISION = 'Jenkinsfile@20191122-02'
```
Comment (stdweird) [inline lib/vsc/install/ci.py]: still don't understand what this is supposed to fix, but at leat the part before the `@` should be form previous constant

File: lib/vsc/install/ci.py [hunk 4]
```diff
@@ -0,0 +1,185 @@
...
+TOX_INI = 'tox.ini'
+TOX_INI_REVISION = 'tox.ini@20191122-01'
```
Comment (stdweird) [inline lib/vsc/install/ci.py]: same remark

File: lib/vsc/install/ci.py [hunk 5]
```diff
@@ -0,0 +1,185 @@
...
+
+logging.basicConfig(format="%(message)s", level=logging.INFO)
+LOG = logging.getLogger()
```
Comment (stdweird) [inline lib/vsc/install/ci.py]: why is this needed? `logging....` should work after previous line

File: lib/vsc/install/ci.py [hunk 6]
```diff
@@ -0,0 +1,185 @@
...
+
+
+def parse_options(args):
+    """Parse options."""
+    parser = OptionParser()
+
+    parser.add_option('-f', '--force', dest='force', action='store_true', help="Use force to overwrite existing files")
```
Comment (stdweird) [inline lib/vsc/install/ci.py]: why is this optional? you are always supposed to use it....

File: lib/vsc/install/ci.py [hunk 7]
```diff
@@ -0,0 +1,185 @@
...
+    lines.extend([
+        "commands = python setup.py test",
+        '',
+        # allow failing tests in Python 3, for now...
+        '[testenv:%s]' % py3_env,
+        "ignore_outcome = true"
+    ])
+
+    txt = '\n'.join(lines)
+    write_file(tox_ini, txt, force=force)
```
Comment (stdweird) [inline lib/vsc/install/ci.py]: do the write in the main and return text here, it will be sooooo much easier to unittest

File: lib/vsc/install/ci.py [hunk 8]
```diff
@@ -0,0 +1,185 @@
...
+    cwd = os.getcwd()
+    jenkinsfile = os.path.join(cwd, JENKINSFILE)
+
+    def indent(line, level=1):
+        """Indent string value with level*4 spaces."""
+        return ' ' * 4 * level + line
+
+    test_cmds = [
+        # make very sure Python 2.7 is available,
+        # since tox ignores failures due to missing Python versions
```
Comment (stdweird) [inline lib/vsc/install/ci.py]: well, since we configured tox to to ignore this (add a ref to the config entry above)

File: lib/vsc/install/ci.py [hunk 9]
```diff
@@ -0,0 +1,185 @@
...
+
+    cwd = os.getcwd()
+    jenkinsfile = os.path.join(cwd, JENKINSFILE)
+
+    def indent(line, level=1):
+        """Indent string value with level*4 spaces."""
+        return ' ' * 4 * level + line
+
+    test_cmds = [
+        # make very sure Python 2.7 is available,
```
Comment (stdweird) [inline lib/vsc/install/ci.py]: so when we start tetsing py36, we either add another version test or we remove the tox ignore missing?

File: lib/vsc/install/ci.py [hunk 10]
```diff
@@ -0,0 +1,185 @@
...
+
+    lines.extend([
+        "commands = python setup.py test",
+        '',
+        # allow failing tests in Python 3, for now...
+        '[testenv:%s]' % py3_env,
+        "ignore_outcome = true"
+    ])
+
+    txt = '\n'.join(lines)
```
Comment (stdweird) [inline lib/vsc/install/ci.py]: add anotehr newline, to always end with newline

File: lib/vsc/install/ci.py [hunk 11]
```diff
@@ -0,0 +1,185 @@
...
+        '',
+        "node {",
+        indent("stage 'checkout git'"),
+        indent("checkout scm"),
+        indent("stage 'test'")
+    ]
+    lines.extend([indent("sh '%s'" % c) for c in test_cmds])
+    lines.append('}')
+
+    txt = '\n'.join(lines)
```
Comment (stdweird) [inline lib/vsc/install/ci.py]: add another newline

File: lib/vsc/install/ci.py [hunk 12]
```diff
@@ -0,0 +1,185 @@
...
+    write_file(jenkinsfile, txt, force=force)
```
Comment (stdweird) [inline lib/vsc/install/ci.py]: do this in main

File: lib/vsc/install/commontest.py [hunk 13]
```diff
@@ -273,3 +275,23 @@ def test_prospector(self):
...
         self.assertFalse(failures, "prospector failures: %s" % pprint.pformat(failures))
+
+    def test_jenkinsfile(self):
+        """Test whether Jenkinsfile is in place, and was auto-generated with correct revision."""
+
+        self.assertTrue(os.path.exists('Jenkinsfile'))
+        with open('Jenkinsfile') as fh:
+            txt = fh.read()
+
+        regex = re.compile(r'^// \[revision: %s\]$' % re.escape(JENKINSFILE_REVISION), re.M)
```
Comment (stdweird) [inline lib/vsc/install/commontest.py]: this is pointless imho. just generate the expected text and compare the generated text with text from file (it gives you a nice diff too).

File: lib/vsc/install/commontest.py [hunk 14]
```diff
@@ -273,3 +275,23 @@ def test_prospector(self):
...
+        self.assertTrue(regex.search(txt), "Pattern '%s' found in tox.ini: %s" % (regex.pattern, txt))
+
+    def test_tox_ini(self):
+        """Test whether tox.ini is in place, and was auto-generated with correct revision."""
+
+        self.assertTrue(os.path.exists('tox.ini'))
+        with open('tox.ini') as fh:
+            txt = fh.read()
+
+        regex = re.compile(r'^# \[revision: %s\]$' % re.escape(TOX_INI_REVISION), re.M)
```
Comment (stdweird) [inline lib/vsc/install/commontest.py]: same remark

File: test/ci.py [hunk 15]
```diff
@@ -0,0 +1,187 @@
...
+    from cStringIO import StringIO  # Python 2
+except ImportError:
+    from io import StringIO  # Python 3
+
+import vsc.install.ci
+from vsc.install.ci import JENKINSFILE_REVISION, TOX_INI_REVISION, gen_jenkinsfile, gen_tox_ini, write_file
+from vsc.install.testing import TestCase
+
+
+def read_file(path):
```
Comment (stdweird) [inline test/ci.py]: why is this a separate function?

File: test/ci.py [hunk 16]
```diff
@@ -0,0 +1,187 @@
...
+class CITest(TestCase):
+    """License related tests"""
+
+    def setUp(self):
+        """Test setup"""
+        super(CITest, self).setUp()
+
+        self.cwd = os.getcwd()
+        self.tmpdir = tempfile.mkdtemp()
+        self.tmpdir_name = os.path.basename(self.tmpdir)
```
Comment (stdweird) [inline test/ci.py]: what is this for?

File: test/ci.py [hunk 17]
```diff
@@ -0,0 +1,187 @@
...
+
+
+class CITest(TestCase):
+    """License related tests"""
+
+    def setUp(self):
+        """Test setup"""
+        super(CITest, self).setUp()
+
+        self.cwd = os.getcwd()
```
Comment (stdweird) [inline test/ci.py]: i'm fine with moving this to `testing.py`, it's used a lot elsewhere too

Comment (stdweird) [reply test/ci.py]: incl next lines

File: test/ci.py [hunk 18]
```diff
@@ -0,0 +1,187 @@
...
+        self.tmpdir = tempfile.mkdtemp()
+        self.tmpdir_name = os.path.basename(self.tmpdir)
+
+    def tearDown(self):
+        """Test cleanup"""
+        shutil.rmtree(self.tmpdir)
```
Comment (stdweird) [inline test/ci.py]: and also this and next line

File: test/ci.py [hunk 19]
```diff
@@ -0,0 +1,187 @@
...
+        os.chdir(self.cwd)
+
+        super(CITest, self).tearDown()
+
+    def run_function(self, function, *args, **kwargs):
```
Comment (stdweird) [inline test/ci.py]: wtf is this? generate the expected text and compare it with a static file in test/data
if you want to test `write_file`, just use mock?

File: test/ci.py [hunk 20]
```diff
@@ -0,0 +1,187 @@
...
+
+        function(*args, **kwargs)
+
+        stdout = stringio.getvalue()
+        vsc.install.ci.LOG.handlers = orig_handlers
+
+        return stdout
+
+    def test_gen_jenkinsfile(self):
+        """Test generating of Jenkinsfile."""
```
Comment (stdweird) [inline test/ci.py]: please make this simpler.

File: Jenkinsfile [hunk 21]
```diff
@@ -1,6 +1,13 @@
...
+// This file was automatically generated using 'python -m vsc.install.ci -f'
+// DO NOT EDIT MANUALLY
+
 node {
-  stage 'checkout git'
-  checkout scm
-  stage 'test'
-  sh "python setup.py test"
-}
+    stage 'checkout git'
```
Comment (boegel) [inline Jenkinsfile]: Note to self: use new syntax here (see https://github.com/hpcugent/quattor/blob/master/Jenkinsfile)

(thx @wdpypere)

File: Jenkinsfile (hunk 1 above)
Comment (boegel) [reply Jenkinsfile]: To be very sure that we're still testing with Python 2.7, since we're configuring `tox` to ignore missing interpreters (so we can get this merged on `jenkins8` where Python 3.6 is not available yet.

File: lib/vsc/install/ci.py (hunk 6 above)
Comment (boegel) [reply lib/vsc/install/ci.py]: Because I really don't like blindly overwriting files by default...
Usually the `Jenkinsfile` and `tox.ini` will be in a git repo, but not always (although I admit it's unlikely).
If you think this is pointless, then I can drop this option and just blindly overwrite.

In practice you'll always run `python -m vsc.install.ci -f`, indeed.

File: lib/vsc/install/ci.py (hunk 9 above)
Comment (boegel) [reply lib/vsc/install/ci.py]: Once we start running the tests on a Jenkins setup that has both Python 2.7 and 3.6, we can drop the `skip_missing_interpreters = True`.

File: test/ci.py (hunk 15 above)
Comment (boegel) [reply test/ci.py]: Just to avoid copy-pasting the same lines in multiple places.

File: test/ci.py (hunk 19 above)
Comment (boegel) [reply test/ci.py]: This is mostly about capturing the logging to stdout (the stuff the user gets to see, at least)

File: test/ci.py (hunk 20 above)
Comment (boegel) [reply test/ci.py]: Please clarify what you mean. ;)

I guess moving the `write_file` to `main` will help quite a bit already...

File: lib/vsc/install/ci.py (hunk 6 above)
Comment (stdweird) [reply lib/vsc/install/ci.py]: i think this is pointless because it is meant for git repos, so any accidental overwrites are still not that bad

File: test/ci.py (hunk 15 above)
Comment (stdweird) [reply test/ci.py]: `read_file(filename)` vs `open(filename).read()`? anyway, try to avoid testing with files. you can typically mock everything you need

File: test/ci.py (hunk 19 above)
Comment (stdweird) [reply test/ci.py]: euh, not sure what he added value of unittesting the log output is, but whatever i guess

File: test/ci.py (hunk 20 above)
Comment (stdweird) [reply test/ci.py]: it's a lot of code to compare 1 static piece of text with a generated one.

File: lib/vsc/install/ci.py (hunk 3 above)
Comment (boegel) [reply lib/vsc/install/ci.py]: I removed it

File: lib/vsc/install/ci.py (hunk 6 above)
Comment (boegel) [reply lib/vsc/install/ci.py]: dropped the `--force`, blindly overwriting now...

File: lib/vsc/install/ci.py (hunk 7 above)
Comment (boegel) [reply lib/vsc/install/ci.py]: done

File: lib/vsc/install/ci.py (hunk 8 above)
Comment (boegel) [reply lib/vsc/install/ci.py]: fixed

File: lib/vsc/install/ci.py (hunk 10 above)
Comment (boegel) [reply lib/vsc/install/ci.py]: fixed

File: lib/vsc/install/ci.py (hunk 12 above)
Comment (boegel) [reply lib/vsc/install/ci.py]: done

File: lib/vsc/install/commontest.py (hunk 13 above)
Comment (boegel) [reply lib/vsc/install/commontest.py]: removed

File: lib/vsc/install/commontest.py (hunk 14 above)
Comment (boegel) [reply lib/vsc/install/commontest.py]: removed

File: test/ci.py (hunk 15 above)
Comment (boegel) [reply test/ci.py]: removed, no longer needed

File: test/ci.py (hunk 17 above)
Comment (boegel) [reply test/ci.py]: removed, no longer needed, but this could indeed be added in `vsc.install.testing` if it pops up again

File: test/ci.py (hunk 19 above)
Comment (boegel) [reply test/ci.py]: removed

File: test/ci.py (hunk 20 above)
Comment (boegel) [reply test/ci.py]: should be a lot better now...

File: Jenkinsfile (hunk 21 above)
Comment (boegel) [reply Jenkinsfile]: fixed in cdd0740

File: lib/vsc/install/ci.py [hunk 22]
```diff
@@ -0,0 +1,163 @@
...
+
+
+def write_file(path, txt):
+    """Write specified contents to specified path."""
+    try:
+        with open(path, 'w') as handle:
+            handle.write(txt)
+        logging.info("Wrote %s", path)
+    except (IOError, OSError) as err:
+        raise IOError("Failed to write %s in %s: %s" % (path, os.getcwd(), err))
```
Comment (stdweird) [inline lib/vsc/install/ci.py]: why is the `in %s` referrring to cwd? if path is relative, prerfix it with cwd before writing

File: lib/vsc/install/ci.py [hunk 23]
```diff
@@ -0,0 +1,163 @@
...
+        # python2.7 must always be available though, see Jenkinsfile
+        "skip_missing_interpreters = true",
+        '',
+        '[testenv]',
+    ]
+
+    if install_vsc_install:
+        lines.extend([
+            # use easy_install rather than pip to install vsc-install dependency
+            # (vsc-* packages may not work when installed with pip due to use of namespace package vsc.*)
```
Comment (stdweird) [inline lib/vsc/install/ci.py]: what is this about? why would one need pip or anything like that?

File: lib/vsc/install/commontest.py [hunk 24]
```diff
@@ -273,3 +275,32 @@ def test_prospector(self):
 
         failures = run_prospector(self.setup.REPO_BASE_DIR)
         self.assertFalse(failures, "prospector failures: %s" % pprint.pformat(failures))
+
+    def test_jenkinsfile(self):
+        """Test whether Jenkinsfile is in place, and was auto-generated with correct revision."""
```
Comment (stdweird) [inline lib/vsc/install/commontest.py]: no revision anymore

File: lib/vsc/install/commontest.py [hunk 25]
```diff
@@ -273,3 +275,32 @@ def test_prospector(self):
...
+            current_jenkinsfile = fh.read()
+
+        expected_jenkinsfile = gen_jenkinsfile()
+
+        error_msg = "Contents of %s does not match expected contents, " % JENKINSFILE
+        error_msg += "you should run 'pyton -m vsc.install.ci' again to re-generate %s" % JENKINSFILE
+        self.assertEqual(current_jenkinsfile, expected_jenkinsfile, error_msg)
+
+    def test_tox_ini(self):
+        """Test whether tox.ini is in place, and was auto-generated with correct revision."""
```
Comment (stdweird) [inline lib/vsc/install/commontest.py]: no revision anymore

File: lib/vsc/install/ci.py [hunk 26]
```diff
@@ -0,0 +1,163 @@
...
+            'commands_pre = python -m easy_install -U vsc-install',
```
Comment (stdweird) [inline lib/vsc/install/ci.py]: why not do this for vsc-install too? it will be there anyway

File: lib/vsc/install/ci.py (hunk 23 above)
Comment (stdweird) [reply lib/vsc/install/ci.py]: ok, think i understand. it 's a bit tricky to bootstrap possibly, but other than that, should be ok. i sortoffrogit how this used to work before...

**Comment by stdweird** [comment]
some minor remarks, looks ok overall

File: lib/vsc/install/ci.py (hunk 23 above)
Comment (boegel) [reply lib/vsc/install/ci.py]: `pip` is the recommended tool to install Python packages (and I think `setuptools`/`easy_install` is even deprecated), but we can't use it because of the way we do the `vsc.*` namespace currently (which is not fully supported by `pip` for some bizarre reason)

File: lib/vsc/install/ci.py (hunk 26 above)
Comment (boegel) [reply lib/vsc/install/ci.py]: This installs the latest `vsc-install` from PyPI, which is not the one we want to test.

But I guess the one in the current directory will take precedence anyway, so I guess it doesn't hurt (and it simplifies things a bit).

**Comment by boegel** [comment]
@stdweird OK now?
